### PR TITLE
Rpc refactoring

### PR DESCRIPTION
**Pull request by bas-vk**
This PR is a start of refactoring the rpc package to reusable API's with support for multiple transport mechanisms. It contains:

- [x] restructuring the rpc package
- [x] ipc support
- [x] a new command `console` (`make console`) which allows a user to open a console to a in the background running geth instance. We might consider using a different name since `console` will have a large change of name conflicts, any suggestions?

It is part of multiple PR's but can be merged and is backwards compatible. The current geth console isn't affected and can still be used for features which are not available in the new console.

In future PR's more API's will be added (admin, debug, miner, ...) which eventually make the current geth console obsolete.

### CODE REVIEW

File: rpc/comms/ipc_windows.go [hunk 1]
```diff
@@ -0,0 +1,699 @@
+// +build windows
+
+package comms
+
+import (
```
Comment (tgerring) [inline rpc/comms/ipc_windows.go]: Minor nit: re-run `go fmt` to clean up indenting issues

**Comment by fjl** [comment]
My suggestion is to make console a geth subcommand.

**Comment by obscuren** [comment]
We have some options here 

1. We unify the daemon and the console and split of mist
2. We have `ethereumd`, `geth` & `mist`

1: Might be a better option now; nice easy debugging, one command.
2: In the long wrong splitting of the daemon and the front ends seems like the right approach. The console, like mist, is just another front end, why should the console be any different.

**Comment by tgerring** [comment]
Before we merge massive RPC changes, we should verify it passes at least as many [RPC tests](https://github.com/ethereum/rpc-tests) as the current codebase does.

**Comment by bas-vk** [comment]
@tgerring, I absolutely agree with testing any changes to the current RPC package extensive before merging them. In this PR the current HTTP RPC and console are deliberately not changed to make this PR not too massive and make a smooth transformation possible. If everyone is satisfied with the approach the next step is to refactor the remaining API's and after that refactor the HTTP RPC interface to the new design and run all RPC tests.

@jfl, personally I would separate both commands for the reason obscuren gives. We can add commands to the console to manage the geth process. E.g. add a shortcut to run geth with the exact arguments as last run.

**Comment by tgerring** [comment]
I ran the RPC Tests against both develop HEAD and this PR branch. Both results are the same:

```
495 passing (2m)
  16 failing
```

There are 5 eth errors and 11 shh errors. Of these, there are 2 nil pointer crashes.

**Comment by fjl** [comment]
I'd still suggest baking the console into `cmd/geth` for now simply because we are shipping one binary
at this time and a lot of external stuff would need to change to adapt to having two. I understand the concern about keeping things neatly separated, but would still argue that it'd be easier for people.

**Comment by obscuren** [comment]
Following errors when running the `console`
```
Error: net_listening method not implemented (-32601)
Error: net_peerCount method not implemented (-32601)
{
  listening: Error: Invalid JSON RPC response,
  getListening: [Function],
  peerCount: Error: Invalid JSON RPC response,
  getPeerCount: [Function]
}
```

```
  version: {
    api: '0.5.0',
    client: 'Geth/v0.9.27/darwin/go1.4',
    getClient: [Function],
    network: Error: Invalid JSON RPC response,
    getNetwork: [Function],
    ethereum: '60',
    getEthereum: [Function],
    whisper: Error: Invalid JSON RPC response,
    getWhisper: [Function]
  },
```

**Pull request closed without merging**